OSD: Deleting transaction object right after applying transaction

somnathr opened the pull request:
Presently, the transaction object is been deleted by the Finisher
thread asynchronously. In heavy load scenario specially if we
unleash journal throttle more, we are seeing high memory usage by
the OSDs because of this. In this new scheme, with the help of
move semantics transaction objects will be deleted synchronously
from the filestore worker threads. We are seeing very much
controllable memory growth now as well as ~3 to 4% cpu usage
benefit because of some reduction of 'new' , 'delete' calls.

Signed-off-by: Somnath Roy <somnath.roy@sandisk.com>

somnathr commented:
Will attach Teuthology report soon , taking ~5-6 days to complete rados tests..

cbodley commented on src/os/ObjectStore.h:
```diff
@@ -1716,71 +1716,53 @@ class ObjectStore {
-	                      NULL, NULL, TrackedOpRef(), handle);
+    vector<Transaction> tls;
+    tls.push_back(std::move(*t));
+    delete t; 
```
the `queue_transaction_and_cleanup()` function should probably go away, since cleanup is now the default behavior of `queue_transactions()`

somnathr replied:
Will do..

cbodley commented on src/os/ObjectStore.h:
```diff
@@ -1716,71 +1716,53 @@ class ObjectStore {
 				TrackedOpRef op = TrackedOpRef(),
 				ThreadPool::TPHandle *handle = NULL) {
-    list<Transaction*> tls;
+    vector<Transaction> tls;
```
the next line should be `tls.push_back(std::move(t));`

somnathr replied:
Why we need to move it again ? 
t is already rval and corresponding push_back(&&) should be called automatically , isn't it ?

cbodley replied:
functions that take rvalue references are a little tricky. they expect the caller to pass an rvalue, but in the body of the function they're treated as an lvalue. so you still need to use std::move() to pass it on as an rvalue

i found this article, [Is an Rvalue Reference an Rvalue?](http://thbecker.net/articles/rvalue_references/section_05.html), to be helpful - it suggests as a rule of thumb, "if it has a name, then it is an lvalue"

somnathr replied:
hmm..Thanks will fix that..

cbodley commented on src/os/ObjectStore.h:
```diff
@@ -493,12 +493,12 @@ class ObjectStore {
-	on_applied_sync.splice(on_applied_sync.end(), (*i)->on_applied_sync);
+	on_applied.splice(on_applied.end(), (*i).on_applied);
+	on_commit.splice(on_commit.end(), (*i).on_commit);
+	on_applied_sync.splice(on_applied_sync.end(), (*i).on_applied_sync);
```
`(*i).` can be simplified to `i->` (here and elsewhere in the patch)

tchaikov replied:
could also use the range based loop to save the pain of `->`.
```c++
for (auto txn : t) {
  on_applied.splice(on_applied.end(), txn.on_applied);
  // ...
}
```

cbodley replied:
that too :) but be careful, you want `auto& txn`, or txn will be copied

tchaikov replied:
oh, right. thanks for pointing this out. should be "auto&"

cbodley commented on src/osd/ReplicatedPG.cc:
```diff
@@ -10180,13 +10179,14 @@ int ReplicatedPG::recover_primary(int max, ThreadPool::TPHandle &handle)
 					      get_osdmap()->get_epoch(),
 					      info.last_complete),
 					    new C_OSD_OndiskWriteUnlock(obc));
+              delete t;
```
we should be able to avoid using new/delete entirely with this new interface. so `t` here should be on the stack, instead of heap-allocated above

somnathr replied:
Will do that..

cbodley commented on src/osd/ECBackend.cc:
```diff
@@ -885,15 +884,11 @@ void ECBackend::handle_sub_write(
+  tls.push_back(std::move(*localt));
+  tls.push_back(std::move(op.t));
   get_parent()->queue_transactions(tls, msg);
+  delete localt;
```
`localt` can be on the stack in this function to avoid new/delete

somnathr replied:
Yes, will do that

cbodley commented on src/osd/OSD.cc:
```diff
@@ -4341,14 +4340,14 @@ void OSD::RemoveWQ::_process(
+    new ContainerContext<PGRef>(pg),
     TrackedOpRef());
 
+  delete t;
```
`t` can be on the stack in this function to avoid new/delete

somnathr replied:
will do that

cbodley commented on src/osd/OSD.cc:
```diff
@@ -6606,9 +6603,10 @@ void OSD::handle_osd_map(MOSDMap *m)
+    std::move(*_t),
+    new C_OnMapApply(&service, pinned_maps, osdmap->get_epoch()),
     0, 0);
+  delete _t;
```
it looks like `_t` could be on the stack, and this funny business above could go away:

```
ObjectStore::Transaction *_t = new ObjectStore::Transaction;
ObjectStore::Transaction &t = *_t;
```

cbodley commented on src/osd/OSD.cc:
```diff
@@ -7242,11 +7240,11 @@ void OSD::dispatch_context_transaction(PG::RecoveryCtx &ctx, PG *pg,
-      ctx.transaction, ctx.on_applied, ctx.on_safe, NULL,
+      std::move(*ctx.transaction), ctx.on_applied, ctx.on_safe, NULL,
       TrackedOpRef(), handle);
+    delete (ctx.transaction);
```
`RecoveryCtx::transaction` should be a non-pointer member to avoid the separate allocation, if possible

somnathr replied:
This needs some changes and in fact we can do this kind of optimization lot of other places specially those are in io/recovery path..
IMO, we should stage it as already the delta of changes for this is quite big, what do you say ?
Also, we need some soak time to see it is working without any problem in al the cases..

cbodley replied:
i agree that the recovery path is less interesting, as long as we cover the main io path

somnathr replied:
Second thought, I am doing it for this and for ECBackend..Probably, it's good to change it now rather than later..Recovery io is also very important for our deployment (since each box is 512 TB) and the memory issues are more likely to hit in the recovery path..

cbodley commented:
The interface looks great! :+1: I commented on a few of the places where callers could use it to avoid extra allocations. We should be able to replace all instances of `new ObjectStore::Transaction` eventually.

liewegas commented on src/os/ObjectStore.cc:
```diff
@@ -146,8 +146,13 @@ ostream& operator<<(ostream& out, const ObjectStore::Sequencer& s)
 
+ostream& operator<<(ostream& out, const ObjectStore::Transaction& tx) {
+
+  return out << "Transaction address = " << &tx;
```
how about

out << "Transaction(" << &tx << ")"

somnathr replied:
Will do..

cbodley replied:
note that the Transaction pointer will change every time it  gets moved, so this won't help us trace a given transaction all the way through the system

somnathr replied:
Yes, I kept it like this since the old code was also printing transaction address only..

liewegas commented:
Cool.  Is nothing needed to make std::move(t) do the right thing when moving a Transaction around?  (I'm not very familiar with how this works...)

Can you also rebase when you address Casey's comments?  Best to get this change in quickly to avoid conflicts.  Thanks!

yuyuyu101 commented:
I guess we need to impl Transaction(Transaction&&) construction, maybe we could use a unique_ptr to point all actual data and make move simpler

somnathr commented:
@cbodley Thanks will address your comments..
@liewegas Yes, we don't have to do anything on Transaction class as it is able to generate implicit move constructor since it is satisfying the following..

there are no user-declared copy constructors;
there are no user-declared copy assignment operators;
there are no user-declared move assignment operators;
there are no user-declared destructors;

somnathr commented:
@yuyuyu101 I don't think we need to implement move constructor for the Transaction class, implicit move constructor should be fine..move constructor for the Buffer class is already there...

yuyuyu101 commented:
struct TransactionImpl{
      TransactionData data;

    void *osr; // NULL on replay

    bool use_tbl;   //use_tbl for encode/decode
    bufferlist tbl;

    map<coll_t, __le32> coll_index;
    map<ghobject_t, __le32, ghobject_t::BitwiseComparator> object_index;

    __le32 coll_id;
    __le32 object_id;

    bufferlist data_bl;
    bufferlist op_bl;

    bufferptr op_ptr;

    list<Context *> on_applied;
    list<Context *> on_commit;
    list<Context *> on_applied_sync;
}

std::unique_ptr<TransactionImpl> impl;

Transaction(Transaction&&rhs):impl(std::move(rhs.impl)) {}

I'm not a move specialist, from my view, a lot of map,list,bufferptr move should also be payload? A single unique_ptr move should be free totally I guess:-).  Maybe @cbodley could give some ideas?

Anyway I like this PR, COOL!

tchaikov commented:
> maybe we could use a unique_ptr to point all actual data and make move simpler

@yuyuyu101 if we go that way, we can not kill the unnecessary new and delete calls.

> I'm not a move specialist, from my view, a lot of map,list,bufferptr move should also be payload? 

see C++ standard N3337, section 12.8/15, 

> The implicitly-defined copy/move constructor for a non-union class X performs a memberwise copy/move of its bases and members. 

as you mentioned, the most noticeable types that can be benefit from move semantics are `map`, `list`, `bufferlist`. all of them have move constructor already. so the road to this pr has been paved =)

efirs commented:
@tchaikov, @cbodley I agree with @yuyuyu101. Now instead of carrying a pointer around we need to copy  sizeof(Transaction), which is 448 bytes.

somnathr commented:
@cbodley @liewegas I have addressed all the comments , please review..

asheplyakov commented on src/test/bench/testfilestore_backend.cc:
```diff
@@ -57,9 +57,10 @@ void TestFileStoreBackend::write(
   os->queue_transaction(
     osr,
-    t,
+    std::move(*t),
```
This one looks a bit fishy. Making the rvalue reference is a promise that no other references exist, which is not the case here: we pass a pointer to C_DeleteTransWrapper which is going to delete it (depending on luck the object might be already deleted)

cbodley replied:
> Making the rvalue reference is a promise that no other references exist

Not really, no. The call to `queue_transaction()` will move data out of the Transaction at memory address `t` (and it does this synchronously), so there's no race between the move and the deletion of `t`.

But you are correct that the `delete t;` added below is a bug, because `C_DeleteTransWrapper` will already delete it. `t` should just be on the stack here, so that both `delete t;` and the `C_DeleteTransWrapper` are unnecessary.

somnathr replied:
Yeah, thanks..I missed it, will fix..Basically, I searched for delete context that is been there in ObjectStore.h (I removed that)..It seems tests are creating their own delete context here.

asheplyakov commented on src/test/objectstore/FileStoreTracker.cc:
```diff
@@ -77,9 +77,10 @@ void FileStoreTracker::submit_transaction(Transaction &t)
+    0, std::move(*out.t),
     new OnApplied(this, in_flight, out.t),
     new OnCommitted(this, in_flight));
+  delete out.t;
```
Same here

somnathr replied:
Will fix the tests..

cbodley commented on src/os/ObjectStore.h:
```diff
@@ -1716,71 +1716,45 @@ class ObjectStore {
   unsigned apply_transaction(Sequencer *osr, Transaction& t, Context *ondisk=0) {
-    list<Transaction*> tls;
-    tls.push_back(&t);
+    vector<Transaction> tls;
```
I think `apply_transaction()` should take `Transaction&& t` as well. This makes it clear that `t` gets moved, so the caller can reuse the same Transaction in multiple calls to `apply_transaction()` without clearing it in between.

cbodley replied:
The ability to reuse a Transaction after a move got me wondering whether that would actually work with the default compiler-generated move operations. I wrote a unit test to verify that a Transaction is `empty()` after a move, and found that it failed.

I opened a separate pull request at https://github.com/ceph/ceph/pull/7303 that includes this unit test, along with custom move operations for struct TransactionData that allow the test to pass.

somnathr replied:
@cbodley The assumption in this interface was user will not be using the transaction object afterwards...But, agreed that interface should be explicitly saying that it moved..
I will do that , but, I saw it's been used lot of places :-(
Regarding empty after move , that's interesting..
Any idea why it is not able to move TransactionData ?

somnathr replied:
Got it..The implicit move constructor for transaction data is not setting ops to 0 and that's why it is not empty..

somnathr replied:
Guess what , I found one place where transaction object is used after apply_transaction()..

Check recursive_remove_collection()...I will fix..

somnathr replied:
@cbodley I am halfway of changing apply_transaction() and I was thinking should we really want to change the apply_transaction() ??
With the current implementation is almost no overhead , but, if we go with move route , there will be two std::move()..Moving transaction has some overhead (at least copying native types)...Let me know what you think..

cbodley replied:
@somnathr there will be one more call to `std::move()`, yes, but all that `std::move()` actually does is cast to an rvalue. the actual move-construct will still only happen once, inside of `vector::push_back()`

somnathr replied:
@cbodley Yeah, that's true..Wasn't thinking while writing :-)..Move name is so confusing....:-)

cbodley commented on src/os/filestore/JournalingObjectStore.cc:
```diff
@@ -81,10 +81,10 @@ int JournalingObjectStore::journal_replay(uint64_t fs_op_seq)
-      Transaction *t = new Transaction(p);
-      tls.push_back(t);
+      Transaction t (p);
+      tls.push_back(std::move(t));
```
you might consider using `tls.emplace_back(p);` here to combine the construction with the push_back()

somnathr replied:
will do..

cbodley commented on src/osd/ECBackend.cc:
```diff
@@ -874,25 +866,20 @@ void ECBackend::handle_sub_write(
-  tls.back()->swap(op.t);
-  tls.back()->register_on_complete(
-    new ObjectStore::C_DeleteTransaction(tls.back()));
+  vector<ObjectStore::Transaction> tls;
```
consider a `tls.reserve(2);` here to make sure we don't do separate allocations for the two `push_back()` calls below

somnathr replied:
Will do..

cbodley commented on src/osd/ReplicatedBackend.cc:
```diff
@@ -1216,9 +1210,9 @@ void ReplicatedBackend::sub_op_modify_impl(OpRequestRef op)
-  list<ObjectStore::Transaction*> tls;
-  tls.push_back(&(rm->localt));
-  tls.push_back(&(rm->opt));
+  vector<ObjectStore::Transaction> tls;
```
consider a `tls.reserve(2);` here to make sure we don't do separate allocations for the two `push_back()` calls below

somnathr replied:
I have another pull request pending that is removing two transaction object from replication path same thing we did for submit_transaction()..
https://github.com/ceph/ceph/pull/7278
If it is correct, we don't need this..
But, probably as an independent pull request, I should do that (?)

somnathr commented:
@cbodley All done , changed apply_transaction also...

cbodley commented on src/osd/ReplicatedBackend.cc:
```diff
@@ -619,16 +619,14 @@ void ReplicatedBackend::submit_transaction(
+  tls.push_back(std::move(*op_t));
   parent->queue_transactions(tls, op.op);
   delete t;
+  delete op_t;
```
it would be nice to get rid of this allocation from RPGTransaction, especially if this is part of main io path. can you make RPGTransaction::t a non-pointer member? then RPGTransaction::get_transaction() would look like:
```
ObjectStore::Transaction&& get_transaction() { return std::move(t); }
```

somnathr replied:
@cbodley Good catch ! This  simple fix we can do now to get rid of one new/delete from io path..
Changed code , running 'make check' will update the pull request once succeeded..

somnathr commented:
@cbodley  @liewegas All done, now all the transaction objects are allocated from stacks..

cbodley commented:
@somnathr looks good to me, great work! have you had a chance to benchmark again?

somnathr commented:
@cbodley Thanks for your review, I am in process of merging all of my write path related commits in a branch and start another bench-marking..Will keep you posted..

somnathr commented:
@cbodley @liewegas Some performance comparison combining all the write path changes..
Setup : 48 OSDS , each OSD 8 TB SSD, Total 384 TB. 1 fio-librbd client with 10 jobs each with QD 128.
3 way replication.

Old code with move constructor but *not all* the transaction object from the stack (The way I submitted pull request originally) , rest of the write path changes are same.


Jobs: 4 (f=4): [_(4),w(3),_(2),w(1)] [14.3% done] [0KB/299.2MB/0KB /s] [0/76.6K/0 iops] [eta 01h:30m:00s]
rbd_iodepth32: (groupid=0, jobs=10): err= 0: pid=625787: Wed Jan 27 10:57:04 2016
  write: io=280055MB, bw=318608KB/s, iops=79651, runt=900093msec
    slat (usec): min=0, max=5256, avg= 3.26, stdev= 7.86
    clat (usec): min=720, max=307960, avg=16056.82, stdev=23429.38
     lat (usec): min=724, max=307961, avg=16060.09, stdev=23429.36
    clat percentiles (usec):
     |  1.00th=[ 1544],  5.00th=[ 2064], 10.00th=[ 2544], 20.00th=[ 3408],
     | 30.00th=[ 4384], 40.00th=[ 5664], 50.00th=[ 7392], 60.00th=[ 9792],
     | 70.00th=[13248], 80.00th=[20096], 90.00th=[40192], 95.00th=[68096],
     | 99.00th=[120320], 99.50th=[136192], 99.90th=[162816], 99.95th=[171008],
     | 99.99th=[189440]



Now, addressing all the comments from cbodley and keeping rest of the write path changes similar, here is the numbers..


Jobs: 3 (f=3): [w(1),_(4),w(2),_(3)] [12.5% done] [0KB/297.3MB/0KB /s] [0/76.2K/0 iops] [eta 01h:45m:00s]
rbd_iodepth32: (groupid=0, jobs=10): err= 0: pid=627208: Wed Jan 27 11:18:45 2016
  write: io=301505MB, bw=343007KB/s, iops=85751, runt=900100msec
    slat (usec): min=0, max=5982, avg= 2.98, stdev= 9.46
    clat (usec): min=628, max=394999, avg=14912.90, stdev=20554.19
     lat (usec): min=631, max=395001, avg=14915.88, stdev=20554.17
    clat percentiles (usec):
     |  1.00th=[ 1464],  5.00th=[ 1992], 10.00th=[ 2448], 20.00th=[ 3344],
     | 30.00th=[ 4320], 40.00th=[ 5664], 50.00th=[ 7392], 60.00th=[ 9792],
     | 70.00th=[13376], 80.00th=[20096], 90.00th=[37120], 95.00th=[57600],
     | 99.00th=[102912], 99.50th=[119296], 99.90th=[156672], 99.95th=[177152],
     | 99.99th=[276480]


See ~7.6% iops improvement gaining 2-3% cpu back (obviously cpu gain is more as some cycles gone to produce extra iops as well) , avg latency is ~7% lower and 99th percentile latency is ~14% lower

mattbenjamin commented:
woot!

liewegas commented:
Can you rebase?  minor conflict with master.  Then I can feed through the next testing round.  Thanks!

somnathr commented:
@liewegas Done..

liewegas commented:
wip-sage-testing3, scheduled at http://pulpito.ceph.com/sage-2016-01-28_12:50:51-rados-wip-sage-testing3---basic-mira
